programs/locker/src/instructions/v3/claim.rs

```diff
@@ -0,0 +1,224 @@
+}
+
+impl ClaimV3Params {
+    pub fn get_total_locked_amount(&self) -> Result<u64> {
+        let total_locked_amount = self
+            .cliff_unlock_amount
+            .safe_add(self.amount_per_period.safe_mul(self.number_of_period)?)?;
+        Ok(total_locked_amount)
+    }
+
+    pub fn verify_recipient(&self, recipient: Pubkey, root: [u8; 32]) -> Result<()> {
+        let total_locked_amount = self.get_total_locked_amount()?;
+        let node = hashv(&[
+            &recipient.key().to_bytes(),
+            &total_locked_amount.to_le_bytes(),
+            &self.vesting_start_time.to_le_bytes(),
+            &self.cliff_time.to_le_bytes(),
+            &self.frequency.to_le_bytes(),
+        ]);
+
+        let leaf = hashv(&[LEAF_PREFIX, &node.to_bytes()]);
+
+        let mut computed_hash = leaf.to_bytes();
+        for p in self.proof.iter() {
+            if computed_hash <= *p {
+                computed_hash = hashv(&[&INTERMEDIATE_PREFIX, &computed_hash, p]).to_bytes();
+            } else {
+                computed_hash = hashv(&[&INTERMEDIATE_PREFIX, p, &computed_hash]).to_bytes();
+            }
+        }
+
+        require!(computed_hash == root, LockerError::InvalidMerkleProof);
+
+        Ok(())
+    }
+
+    pub fn get_max_unlocked_amount(&self, current_ts: u64) -> Result<u64> {
+        if current_ts < self.cliff_time {
+            return Ok(0);
+        }
+        let period = current_ts
+            .safe_sub(self.cliff_time)?
+            .safe_div(self.frequency)?;
+        let period = period.min(self.number_of_period);
+
+        let unlocked_amount = self
+            .cliff_unlock_amount
+            .safe_add(period.safe_mul(self.amount_per_period)?)?;
+
+        Ok(unlocked_amount)
+    }
+
+    pub fn get_claimable_amount(&self, total_claimed_amount: u64, current_ts: u64) -> Result<u64> {
+        let max_unlocked_amount = self.get_max_unlocked_amount(current_ts)?;
+        let claimable_amount = max_unlocked_amount.safe_sub(total_claimed_amount)?;
+        Ok(claimable_amount)
+    }
+
+    pub fn get_claim_amount(&self, total_claimed_amount: u64) -> Result<u64> {
+        let current_ts = Clock::get()?.unix_timestamp as u64;
+        let claimable_amount = self.get_claimable_amount(total_claimed_amount, current_ts)?;
+
+        let amount = claimable_amount.min(self.max_amount);
+
+        Ok(amount)
+    }
+
+    pub fn is_recipient_claimed_full_amount(&self, total_claimed_amount: u64) -> Result<bool> {
+        let max_amount = self.get_total_locked_amount()?;
+        Ok(total_claimed_amount == max_amount)
+    }
+}
+
+#[event_cpi]
+#[derive(Accounts)]
+pub struct ClaimV3<'info> {
+    /// Claim status PDA
+    #[account(
+        init_if_needed,
+        seeds = [
+            b"claim_status".as_ref(),
+            recipient.key().to_bytes().as_ref(),
+            escrow.key().to_bytes().as_ref()
+        ],
+        bump,
+        space = 8 + ClaimStatus::INIT_SPACE,
+        payer = recipient
```

Comment:
we should use other payer to pay rent fee

Reply:
This small account has a one-time create, I’ve included it here for convenience.
Otherwise, we’d need a separate instruction.

Reply:
The probably is creator may use script to claim, so only 1 address has SOL, and all recipients are only used for sign, but not pay rent fee

Reply:
So, solution will be:
- Create a separate instruction for claim_status.
- At client side will push instruction create claim_status if not exited in claim transaction.
But if script have many instruction will reach limit of transaction size.

Reply:
No, you can just use payer as seprated signer from recipient is fine

Reply:
i fixed it